### PR TITLE
Add vector type to Dgraph

### PR DESCRIPTION
**Pull request by harshil-goel**

### CODE REVIEW

File: types/value.go [hunk 1]
```diff
@@ -17,10 +17,86 @@
...
+	for i := 0; i < len(v); i++ {
+		bits := math.Float32bits(v[i])
+		binary.LittleEndian.PutUint32(offset, bits)
+		offset = offset[4:]
+	}
+	return retVal
+}
+
+func FloatArrayAsString(v []float32) string {
+	retVal := "["
```
Comment (mangalaman93) [inline types/value.go:89]: we should strings.Builder here

**Review by mangalaman93 — CHANGES_REQUESTED** [review]
We should check for coverage for the new changes using unit tests

File: types/conversion.go [hunk 2]
```diff
@@ -33,6 +34,73 @@ import (
...
+// according to the following ebnf:
+//
+//	floatArray ::= "[" [floatList] [whitespace] "]"
+//	floatList := float32Val |
+//	             float32Val floatSpaceList |
+//	             float32Val floatCommaList
+//	floatSpaceList := (whitespace float32Val)+
+//	floatCommaList := ([whitespace] "," [whitespace] float32Val)+
+//	float32Val := < a string rep of a float32 value >
+func ParseVFloat(s string) ([]float32, error) {
```
[the remark this answers is not in the document]
Comment (harshil-goel) [reply types/conversion.go:48]: We don't have lexer for an array.

File: query/outputnode.go [hunk 3]
```diff
@@ -1560,7 +1574,7 @@ func (sg *SubGraph) preTraverse(enc *encoder, uid uint64, dst fastJsonNode) erro
 					if lang != "" && lang != "*" {
 						fieldNameWithTag += "@" + lang
 					}
-					encodeAsList := pc.List && lang == ""
+					encodeAsList := (pc.List && lang == "")
```
Comment (mangalaman93) [inline query/outputnode.go:1577]: we don't need this change

File: go.mod [hunk 4]
```diff
@@ -15,7 +15,7 @@ require (
 	github.com/Masterminds/semver/v3 v3.1.0
 	github.com/blevesearch/bleve/v2 v2.3.10
 	github.com/dgraph-io/badger/v4 v4.2.0
-	github.com/dgraph-io/dgo/v230 v230.0.1
+	github.com/dgraph-io/dgo/v230 v230.0.2-0.20230929195431-42a4d4ef28da
```
Comment (mangalaman93) [inline go.mod:18]: we need to merge dgo PR first

File: types/value.go [hunk 5]
```diff
@@ -17,10 +17,61 @@
...
+	"unsafe"
 )
 
+// BytesAsFloatArray(encoded) converts encoded into a []float32.
+// If len(encoded) % 4 is not 0, it will ignore any trailing
+// bytes, and simply convert 4 bytes at a time to generate the
+// float64 entries.
+// Current implementation assuming littleEndian encoding
+func BytesAsFloatArray(encoded []byte) []float32 {
+
```
Comment (mangalaman93) [inline types/value.go:33]: nit: extra newline

File: query/query0_test.go [hunk 6]
```diff
@@ -31,6 +31,62 @@ import (
...
+				a as vectorNonIndex
+			}
+			aggregation() {
+				avg(val(a))
+				sum(val(a))
+			}
+		}
+	`
+	js := processQueryNoErr(t, query)
+	k := `{
```
Comment (mangalaman93) [inline query/query0_test.go:47]: can we format the query better so it is readable.

File: query/query0_test.go [hunk 7]
```diff
@@ -31,6 +31,62 @@ import (
...
+  "data": {
+    "me": [
+      {
+        "vectorNonIndex": [
```
Comment (mangalaman93) [inline query/query0_test.go:51]: nit: you could put these in one line

File: types/conversion.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mangalaman93) [reply types/conversion.go:48]: Let's put a todo here for now

File: types/scalar_types.go [hunk 8]
```diff
@@ -57,21 +57,28 @@ const (
...
+	"binary":        BinaryID,
+	"int":           IntID,
+	"float":         FloatID,
+	"bool":          BoolID,
+	"datetime":      DateTimeID,
+	"geo":           GeoID,
+	"uid":           UidID,
+	"string":        StringID,
+	"password":      PasswordID,
+	"float32vector": VFloatID,
```
Comment (mangalaman93) [inline types/scalar_types.go:81]: check for the name once with the leadership

**Review by mangalaman93 — CHANGES_REQUESTED** [review]

**Review by mangalaman93 — DISMISSED** [review]

**Review by mangalaman93 — APPROVED** [review]

**Pull request merged**